File: LRUCache.java
// Data structures to synchronize: HashMap, doubly LinkedList's head and tail
class ListNode<K, V> {
  K key;
  V value;
  ListNode<K, V> prev, next;
  
  public ListNode(K key, V value) {
    this.key = key;
    this.value = value;
    prev = next = null;
  }
}

public class LRUCache<K, V> {
  ListNode<K, V> head, tail;
  Map<K, ListNode> map;
  int limit;
  
  // limit is the max capacity of the cache
  public LRUCache(int limit) {
    head = tail = null;
    map = new HashMap<>();
    this.limit = limit;    
  }
  
  // If key exists, reset it to the new value and move the node to the tail
  // Otherwise, create a new node and append to tail. If the limit is exceeded, delete the head
  public void set(K key, V value) {
    ListNode<K, V> curr = map.get(key);
    if (curr != null) {
      curr.value = value;
      moveExistingToTail(curr);
    } else {
      curr = new ListNode<>(key, value);
      if (tail == null) {
        head = tail = curr;
      } else {
        tail.next = curr;
        curr.prev = tail;
        tail = curr;
      }
      map.put(key, curr);
      if (map.size() > limit) {
        map.remove(head.key);
        head.next.prev = null;
        head = head.next;
        if (head == null) {
          tail = null;
        }
      } 
    }
  }
  
  public V get(K key) {
    ListNode<K, V> curr = map.get(key);
    if (curr == null) {
      return null;
    } else {
      V res = curr.value;
      moveExistingToTail(curr);
      return res;
    }
  }
  
  private void moveExistingToTail(ListNode<K, V> curr) {
    if (curr == null || curr.next == null) {
      return;
    }
    ListNode<K, V> prev, next;
    prev = curr.prev;
    next = curr.next;
    if (prev != null) {
      prev.next = next;
    } else {
      head = next;
    }
    next.prev = prev;
    tail.next = curr;
    curr.next = null;
    curr.prev = tail;
    tail = tail.next;  
  }
}

File: lru_cache.py
# With dummy nodes
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = self.next = None

class LRUCache:

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self._dic = {}              # key -> node
        self._head = Node(-1, -1)   # Dummy head
        self._tail = Node(-1, -1)   # Dummy tail
        self._head.next, self._tail.prev = self._tail, self._head   # h <-> t
        self._capacity = capacity
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        node = self._dic.get(key)
        if node is None:
            return -1
        self._move_to_tail(node)
        return node.val

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        node = self._dic.get(key)
        if node is not None:
            node.val = value
            self._move_to_tail(node)
        else:
            node = Node(key, value)
            if len(self._dic) == self._capacity:
                self._remove_head()
            self._append_to_tail(node)
            self._dic[key] = node
    
    def _move_to_tail(self, node):
        """Disconnect the node from the doubly linked list and move it to the tail."""
        node_prev, node_next = node.prev, node.next
        node.prev, node.next = None, None
        node_prev.next, node_next.prev = node_next, node_prev
        self._append_to_tail(node)
    
    def _append_to_tail(self, node):
        """Append a free node to the tail of the doubly linked list."""
        tail_prev = self._tail.prev
        tail_prev.next = self._tail.prev = node
        node.prev, node.next = tail_prev, self._tail
   
    def _remove_head(self):
        """Remove the head node and delete it from the hashmap."""
        node, node_next = self._head.next, self._head.next.next
        self._head.next, node_next.prev = node_next, self._head
        self._dic.pop(node.key)

        